utils_test.go

```diff
@@ -42,3 +50,202 @@ func TestValidStoragePathFormat(t *testing.T) {
 		assert.NilError(t, err)
 	}
 }
+
+func TestRootlessRuntimeDir(t *testing.T) {
+	tmpDir, err := ioutil.TempDir(os.TempDir(), "containers-user-*")
+	defer os.RemoveAll(tmpDir)
+	assert.NilError(t, err)
+
+	// testing homedir.GetRuntimeDir binding.
+
+	monkey.Patch(homedir.GetRuntimeDir, func() (string, error) {
+		return tmpDir, nil
+	})
+	defer monkey.Unpatch(homedir.GetRuntimeDir)
+
+	const rootlessUID = math.MaxUint16
+
+	rootlessRuntimeDir, err := getRootlessRuntimeDir(rootlessUID)
+	assert.NilError(t, err)
+	assert.Assert(t, rootlessRuntimeDir == tmpDir, "Rootless runtime dir should follow homedir.GetRuntimeDir.")
+
+	monkey.Patch(homedir.GetRuntimeDir, func() (string, error) {
+		return "", errors.New("homedir.GetRuntimeDir is disabled")
+	})
+
+	// testing systemd dir for systemd based os.
+
+	systemdRuntimeDir := fmt.Sprintf("/run/user/%d", rootlessUID)
+
+	var readFileGuard *monkey.PatchGuard
+	readFileGuard = monkey.Patch(ioutil.ReadFile, func(path string) ([]byte, error) {
+		readFileGuard.Unpatch()
+		defer readFileGuard.Restore()
+		if path == "/proc/1/comm" {
+			return []byte("systemd"), nil
+		}
+		return ioutil.ReadFile(path)
+	})
+	defer monkey.Unpatch(ioutil.ReadFile)
+
+	var lstatGuard *monkey.PatchGuard
+	lstatGuard = monkey.Patch(system.Lstat, func(path string) (*system.StatT, error) {
+		lstatGuard.Unpatch()
+		defer lstatGuard.Restore()
+		if path == systemdRuntimeDir {
+			err = os.Chmod(tmpDir, 0700)
+			assert.NilError(t, err)
+			err = os.Chown(tmpDir, os.Getuid(), os.Getgid())
```

Comment:
This probably does nothing, and if it does something, it won’t work as non-root users; that seems rather problematic for unit tests at least, for any kind of edit+compile+test cycle with an IDE.

@nalind what’s a good way to integrate such tests in this project? Most of the CI runs in dedicated VMs…

Reply:
`tmpDir` was created by current user, so this user can change 2 group and others permissions in any possible way. It shouldn't provide problems.

Why we have to have `tmpDir`? Because `system.StatT` type has all fields private and we can't mock it without using any of 2 public functions. The easiest way is to have `tmpDir` and change its permissions.

Reply:
> `tmpDir` was created by current user, so this user can change 2 group and others permissions in any possible way. It shouldn't provide problems.

I’m not worried about the human running this unit test attacking the operation of the unit test. What would be the point? And such an attacker could just as well change the permissions _after_ this line executes anyway.

> Why we have to have `tmpDir`? Because `system.StatT` type has all fields private and we can't mock it without using any of 2 public functions. The easiest way is to have `tmpDir` and change its permissions.

Sure, that’s fair for the `Chmod`, but if the `Chown` is not really intended to _change_ permissions, the intent would be more clearly expressed by just calling `Lstat` and verifying the ownership has the expected values.

Reply:
Ideally we'd be running the unit tests for filesystem-specific logic along with the integration tests that are geared for it.  In that context, we'd probably have to extend the `local-test-unit` target to also compile tests which can/should be run without privileges with `go test -c`, and then invoke the result using a `-test.run` flag that skipped tests which are expected to fail when they're run without root privileges.

Reply:
> calling Lstat and verifying the ownership has the expected values

Sure, we will fix it. `Chown` call is redundant here.

Reply:
> which can/should be run without privileges

This test should work without priviledges, sure, we can remove `Chown`.